src/MongoDB.Driver/Core/Misc/TaskExtensions.cs

```
@@ -15,12 +15,76 @@
 
 using System;
 using System.Runtime.CompilerServices;
+using System.Threading;
 using System.Threading.Tasks;
 
 namespace MongoDB.Driver.Core.Misc
 {
     internal static class TaskExtensions
     {
+        #if !NET6_0_OR_GREATER
+
+        public static async Task WaitAsync(this Task task, TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            ValidateTimeout(timeout);
```

These extension methods are nice!

But there is a subtle issue here that we may or may not want to address.

The way this is currently coded when passing an invalid timeout the exception will NOT be thrown immediately. Instead a Task will be returned and the exception will be thrown and observed by the code that awaits the Task.

To immediately throw an exception without returning a Task requires some tricks involving an async helper method.

I leave it up to you to decide if you want to be more precise about when the exception is thrown.

---

For one example of how to throw an exception immediately instead of returning a Task see:

https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/[Task.cs](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs,2825),2825

---

Do you mean negative timeout by "an invalid timeout"? If so - it should throw immediately, see the ValidateTimeout method.

---

As per our discussion - it appears you are right. Will split the implementation as per your advice.